File: mac/Sources/PixShell/Monitor/SysInfoParser.swift
import Foundation

/// 系统信息采集文本解析：纯逻辑，无 AppKit 依赖，方便单测。
/// 输入是 `SysInfoPanel.command` 采集回来的 `KEY=value` 文本（表格类字段用 `KEY=a\tb\tc` 重复行）。
/// 任何字段缺失/格式异常都不应崩溃：一律退化为 nil / 空数组。
enum SysInfoParser {

    /// CPU 一栏：型号 + 核数 + 频率/缓存/BogoMIPS + 占用率拆分。
    struct CPUInfo {
        var model: String?
        var cores: Int?
        var mhz: String?
        var cache: String?
        var bogomips: String?
        var busyPct: Double?
        var userPct: Double?
        var systemPct: Double?
        var idlePct: Double?
        var iowaitPct: Double?
    }

    /// 网卡一行：名称 / IPv4 / MAC / 收发字节数。
    struct NetRow {
        var name: String
        var ip: String?
        var mac: String?
        var rxBytes: Int64?
        var txBytes: Int64?
    }

    /// 磁盘一行：挂载点 / 容量 / 已用 / 可用 / 使用率 / 文件系统。
    struct DiskRow {
        var mount: String
        var size: String?
        var used: String?
        var avail: String?
        var pct: Int?
        var fs: String?
    }

    struct SysInfo {
        // 基本
        var hostname: String?
        var distro: String?
        var kernel: String?
        var arch: String?
        var uptime: String?
        var load: String?
        var ip: String?
        // CPU
        var cpu = CPUInfo()
        // 内存 / 交换
        var memPct: Int?
        var memUsedMB: Int?
        var memTotalMB: Int?
        var swapPct: Int?
        var swapUsedMB: Int?
        var swapTotalMB: Int?
        // 表格
        var net: [NetRow] = []
        var disks: [DiskRow] = []
    }

    /// 解析采集文本；未知/缺失/损坏的行一律跳过，永不崩溃。
    static func parse(_ raw: String) -> SysInfo {
        var info = SysInfo()
        for rawLine in raw.split(separator: "\n", omittingEmptySubsequences: true) {
            let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !line.isEmpty else { continue }
            guard let eq = line.firstIndex(of: "=") else { continue }
            let key = String(line[line.startIndex..<eq])
            let value = String(line[line.index(after: eq)...])
            apply(key: key, value: value, into: &info)
        }
        return info
    }

    private static func apply(key: String, value: String, into info: inout SysInfo) {
        switch key {
        case "hostname": info.hostname = nonEmpty(value)
        case "distro": info.distro = nonEmpty(value)
        case "kernel": info.kernel = nonEmpty(value)
        case "arch": info.arch = nonEmpty(value)
        case "uptime": info.uptime = nonEmpty(value)
        case "load": info.load = nonEmpty(value)
        case "ip": info.ip = nonEmpty(value)
        case "cpu_model": info.cpu.model = nonEmpty(value)
        case "cpu_cores": info.cpu.cores = Int(value.trimmingCharacters(in: .whitespaces))
        case "cpu_mhz": info.cpu.mhz = nonEmpty(value)
        case "cpu_cache": info.cpu.cache = nonEmpty(value)
        case "cpu_bogomips": info.cpu.bogomips = nonEmpty(value)
        case "cpu_busy": info.cpu.busyPct = Double(value)
        case "cpu_user": info.cpu.userPct = Double(value)
        case "cpu_system": info.cpu.systemPct = Double(value)
        case "cpu_idle": info.cpu.idlePct = Double(value)
        case "cpu_iowait": info.cpu.iowaitPct = Double(value)
        case "mem_pct": info.memPct = Int(value)
        case "mem_used_mb": info.memUsedMB = Int(value)
        case "mem_total_mb": info.memTotalMB = Int(value)
        case "swap_pct": info.swapPct = Int(value)
        case "swap_used_mb": info.swapUsedMB = Int(value)
        case "swap_total_mb": info.swapTotalMB = Int(value)
        case "net_row":
            let f = value.components(separatedBy: "\t")
            guard let name = f.first, !name.isEmpty else { return }
            let row = NetRow(
                name: name,
                ip: f.count > 1 ? nonEmpty(f[1]) : nil,
                mac: f.count > 2 ? nonEmpty(f[2]) : nil,
                rxBytes: f.count > 3 ? Int64(f[3]) : nil,
                txBytes: f.count > 4 ? Int64(f[4]) : nil
            )
            info.net.append(row)
        case "disk_row":
            let f = value.components(separatedBy: "\t")
            guard let mount = f.first, !mount.isEmpty else { return }
            let pctStr = f.count > 4 ? f[4].replacingOccurrences(of: "%", with: "") : ""
            let row = DiskRow(
                mount: mount,
                size: f.count > 1 ? nonEmpty(f[1]) : nil,
                used: f.count > 2 ? nonEmpty(f[2]) : nil,
                avail: f.count > 3 ? nonEmpty(f[3]) : nil,
                pct: Int(pctStr),
                fs: f.count > 5 ? nonEmpty(f[5]) : nil
            )
            info.disks.append(row)
        default:
            break // 未知 key：忽略，兼容未来新增字段
        }
    }

    private static func nonEmpty(_ s: String) -> String? {
        let t = s.trimmingCharacters(in: .whitespaces)
        return (t.isEmpty || t == "-") ? nil : t
    }

    /// 字节数转人类可读（KB/MB/GB），用于网卡 rx/tx 展示。
    static func formatBytes(_ n: Int64?) -> String {
        guard let n = n, n >= 0 else { return "-" }
        let units = ["B", "KB", "MB", "GB", "TB"]
        var v = Double(n)
        var i = 0
        while v >= 1024, i < units.count - 1 { v /= 1024; i += 1 }
        return i == 0 ? "\(n) B" : String(format: "%.1f %@", v, units[i])
    }
}

File: win/Monitor/SysInfoParser.cs
using System;
using System.Collections.Generic;
using System.Globalization;

namespace PixShell.Monitor;

/// <summary>
/// 系统信息采集文本解析：纯逻辑，无 UI 依赖，字段与解析规则 1:1 对齐 mac 版
/// Monitor/SysInfoParser.swift。输入是 <see cref="PixShell.UI.SysInfoWindow.Command"/>
/// 采集回来的 `KEY=value` 文本（表格类字段用 `KEY=a\tb\tc\td\te` 重复行）。
/// 任何字段缺失/格式异常都不应崩溃：一律退化为 null / 空列表。
/// </summary>
public static class SysInfoParser
{
    /// <summary>CPU 一栏：型号 + 核数 + 频率/缓存/BogoMIPS + 占用率拆分。</summary>
    public sealed class CpuInfo
    {
        public string? Model;
        public int? Cores;
        public string? Mhz;
        public string? Cache;
        public string? Bogomips;
        public double? BusyPct;
        public double? UserPct;
        public double? SystemPct;
        public double? IdlePct;
        public double? IowaitPct;
    }

    /// <summary>网卡一行：名称 / IPv4 / MAC / 收发字节数。</summary>
    public sealed class NetRow
    {
        public string Name = "";
        public string? Ip;
        public string? Mac;
        public long? RxBytes;
        public long? TxBytes;
    }

    /// <summary>磁盘一行：挂载点 / 容量 / 已用 / 可用 / 使用率 / 文件系统。</summary>
    public sealed class DiskRow
    {
        public string Mount = "";
        public string? Size;
        public string? Used;
        public string? Avail;
        public int? Pct;
        public string? Fs;
    }

    public sealed class SysInfo
    {
        // 基本
        public string? Hostname;
        public string? Distro;
        public string? Kernel;
        public string? Arch;
        public string? Uptime;
        public string? Load;
        public string? Ip;
        // CPU
        public CpuInfo Cpu = new();
        // 内存 / 交换
        public int? MemPct;
        public int? MemUsedMb;
        public int? MemTotalMb;
        public int? SwapPct;
        public int? SwapUsedMb;
        public int? SwapTotalMb;
        // 表格
        public List<NetRow> Net = new();
        public List<DiskRow> Disks = new();
    }

    /// <summary>解析采集文本；未知/缺失/损坏的行一律跳过，永不崩溃。</summary>
    public static SysInfo Parse(string raw)
    {
        var info = new SysInfo();
        if (string.IsNullOrEmpty(raw)) return info;
        foreach (var rawLine in raw.Split('\n'))
        {
            var line = rawLine.Trim();
            if (line.Length == 0) continue;
            var eq = line.IndexOf('=');
            if (eq < 0) continue;
            var key = line[..eq];
            var value = line[(eq + 1)..];
            Apply(key, value, info);
        }
        return info;
    }

    private static void Apply(string key, string value, SysInfo info)
    {
        switch (key)
        {
            case "hostname": info.Hostname = NonEmpty(value); break;
            case "distro": info.Distro = NonEmpty(value); break;
            case "kernel": info.Kernel = NonEmpty(value); break;
            case "arch": info.Arch = NonEmpty(value); break;
            case "uptime": info.Uptime = NonEmpty(value); break;
            case "load": info.Load = NonEmpty(value); break;
            case "ip": info.Ip = NonEmpty(value); break;
            case "cpu_model": info.Cpu.Model = NonEmpty(value); break;
            case "cpu_cores": info.Cpu.Cores = ParseInt(value); break;
            case "cpu_mhz": info.Cpu.Mhz = NonEmpty(value); break;
            case "cpu_cache": info.Cpu.Cache = NonEmpty(value); break;
            case "cpu_bogomips": info.Cpu.Bogomips = NonEmpty(value); break;
            case "cpu_busy": info.Cpu.BusyPct = ParseDouble(value); break;
            case "cpu_user": info.Cpu.UserPct = ParseDouble(value); break;
            case "cpu_system": info.Cpu.SystemPct = ParseDouble(value); break;
            case "cpu_idle": info.Cpu.IdlePct = ParseDouble(value); break;
            case "cpu_iowait": info.Cpu.IowaitPct = ParseDouble(value); break;
            case "mem_pct": info.MemPct = ParseInt(value); break;
            case "mem_used_mb": info.MemUsedMb = ParseInt(value); break;
            case "mem_total_mb": info.MemTotalMb = ParseInt(value); break;
            case "swap_pct": info.SwapPct = ParseInt(value); break;
            case "swap_used_mb": info.SwapUsedMb = ParseInt(value); break;
            case "swap_total_mb": info.SwapTotalMb = ParseInt(value); break;
            case "net_row":
            {
                var f = value.Split('\t');
                if (f.Length == 0 || f[0].Length == 0) return;
                info.Net.Add(new NetRow
                {
                    Name = f[0],
                    Ip = f.Length > 1 ? NonEmpty(f[1]) : null,
                    Mac = f.Length > 2 ? NonEmpty(f[2]) : null,
                    RxBytes = f.Length > 3 ? ParseLong(f[3]) : null,
                    TxBytes = f.Length > 4 ? ParseLong(f[4]) : null,
                });
                break;
            }
            case "disk_row":
            {
                var f = value.Split('\t');
                if (f.Length == 0 || f[0].Length == 0) return;
                var pctStr = f.Length > 4 ? f[4].Replace("%", "") : "";
                info.Disks.Add(new DiskRow
                {
                    Mount = f[0],
                    Size = f.Length > 1 ? NonEmpty(f[1]) : null,
                    Used = f.Length > 2 ? NonEmpty(f[2]) : null,
                    Avail = f.Length > 3 ? NonEmpty(f[3]) : null,
                    Pct = ParseInt(pctStr),
                    Fs = f.Length > 5 ? NonEmpty(f[5]) : null,
                });
                break;
            }
            default:
                break; // 未知 key：忽略，兼容未来新增字段
        }
    }

    private static string? NonEmpty(string s)
    {
        var t = s.Trim();
        return (t.Length == 0 || t == "-") ? null : t;
    }

    private static int? ParseInt(string s) =>
        int.TryParse(s.Trim(), NumberStyles.Integer, CultureInfo.InvariantCulture, out var v) ? v : null;

    private static long? ParseLong(string s) =>
        long.TryParse(s.Trim(), NumberStyles.Integer, CultureInfo.InvariantCulture, out var v) ? v : null;

    private static double? ParseDouble(string s) =>
        double.TryParse(s.Trim(), NumberStyles.Float, CultureInfo.InvariantCulture, out var v) ? v : null;

    /// <summary>字节数转人类可读（KB/MB/GB），用于网卡 rx/tx 展示。</summary>
    public static string FormatBytes(long? n)
    {
        if (n is null || n < 0) return "-";
        string[] units = { "B", "KB", "MB", "GB", "TB" };
        double v = n.Value;
        int i = 0;
        while (v >= 1024 && i < units.Length - 1) { v /= 1024; i++; }
        return i == 0 ? $"{n} B" : $"{v.ToString("0.0", CultureInfo.InvariantCulture)} {units[i]}";
    }
}
